src/main/scala/is/hail/expr/ir/functions/UtilFunctions.scala

```diff
@@ -17,23 +17,92 @@ object UtilFunctions extends RegistryFunctions {
+      val one = Literal(1, coerce[TArray](a.typ).elementType)
+      ArrayFold(a, one, product, v, If(IsNA(Ref(v)), Ref(product), ApplyBinaryPrimOp(Multiply(), Ref(product), Ref(v))))
+    }
+
     registerIR("sum", TAggregable(tnum("T")))(ApplyAggOp(_, Sum(), Seq()))
 
+    registerIR("count", TAggregable(tv("T"))) { agg =>
+      val uid = genUID()
+      ApplyAggOp(AggMap(agg, uid, MakeStruct(Seq())), Count(), Seq())
+    }
+
     registerIR("min", TArray(tnum("T"))) { a =>
-      val body = If(IsNA(Ref("min")),
-        Ref("value"),
-        If(IsNA(Ref("value")),
-          Ref("min"),
-          If(ApplyBinaryPrimOp(LT(), Ref("min"), Ref("value")), Ref("min"), Ref("value"))))
-      ArrayFold(a, NA(tnum("T").t), "min", "value", body)
+      val min = genUID()
+      val value = genUID()
+      val body = If(IsNA(Ref(min)),
+        Ref(value),
+        If(IsNA(Ref(value)),
+          Ref(min),
+          If(ApplyBinaryPrimOp(LT(), Ref(value), Ref(min)), Ref(value), Ref(min))))
+      ArrayFold(a, NA(tnum("T").t), min, value, body)
+    }
+
+    registerIR("max", TArray(tnum("T"))) { a =>
+      val max = genUID()
+      val value = genUID()
+      val body = If(IsNA(Ref(max)),
+        Ref(value),
+        If(IsNA(Ref(value)),
+          Ref(max),
+          If(ApplyBinaryPrimOp(GT(), Ref(value), Ref(max)), Ref(value), Ref(max))))
+      ArrayFold(a, NA(tnum("T").t), max, value, body)
     }
 
     registerIR("isDefined", tv("T")) { a => ApplyUnaryPrimOp(Bang(), IsNA(a)) }
+    registerIR("isMissing", tv("T")) { a => IsNA(a) }
 
     registerIR("[]", TArray(tv("T")), TInt32()) { (a, i) => ArrayRef(a, i) }
 
+    registerIR("[:]", TArray(tv("T"))) { (a) => a }
+
+    registerIR("[*:]", TArray(tv("T")), TInt32()) { (a, i) =>
+      val idx = genUID()
+      ArrayMap(
+        ArrayRange(If(ApplyBinaryPrimOp(LT(), i, I32(0)),
+          ApplyBinaryPrimOp(Add(), ArrayLen(a), i),
+          i),
+          ArrayLen(a),
+          I32(1)),
+        idx,
+        ArrayRef(a, Ref(idx)))
+    }
+
+    registerIR("[:*]", TArray(tv("T")), TInt32()) { (a, i) =>
+      val idx = genUID()
+      ArrayMap(
+        ArrayRange(I32(0),
+          If(ApplyBinaryPrimOp(LT(), i, I32(0)),
+            ApplyBinaryPrimOp(Add(), ArrayLen(a), i),
+            i),
+          I32(1)),
+        idx,
+        ArrayRef(a, Ref(idx)))
+    }
+
+    registerIR("[*:*]", TArray(tv("T")), TInt32(), TInt32()) { (a, i, j) =>
+      val idx = genUID()
+      ArrayMap(
+        ArrayRange(
+          If(ApplyBinaryPrimOp(LT(), i, I32(0)),
+            ApplyBinaryPrimOp(Add(), ArrayLen(a), i),
+            i),
+          If(ApplyBinaryPrimOp(LT(), j, I32(0)),
+            ApplyBinaryPrimOp(Add(), ArrayLen(a), j),
+            j),
+          I32(1)),
+        idx,
+        ArrayRef(a, Ref(idx)))
+    }
+
     registerIR("[]", tv("T", _.isInstanceOf[TTuple]), TInt32()) { (a, i) => GetTupleElement(a, i.asInstanceOf[I32].x) }
 
+    registerIR("[:]", TString()) { (a) => a }
```

Comment:
Do we still need something like StringRange to add the other slicing functions here?

Reply:
No, we just need to convert the strings to java.lang.String, but that requires more work.